client/client.go

```diff
@@ -0,0 +1,368 @@
+package client
+
+import (
+	"context"
+	"math/big"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/mysteriumnetwork/payments/bindings"
+	"github.com/mysteriumnetwork/payments/crypto"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+// ProviderChannel represents the provider channel
+type ProviderChannel struct {
+	Beneficiary   common.Address
+	Balance       *big.Int
+	Settled       *big.Int
+	Loan          *big.Int
+	LastUsedNonce *big.Int
+	Timelock      *big.Int
+}
+
+// Blockchain contains all the useful blockchain utilities for the payment off chain messaging
+type Blockchain struct {
+	client    *ethclient.Client
+	bcTimeout time.Duration
+}
+
+// NewBlockchain returns a new instance of blockchain
+func NewBlockchain(c *ethclient.Client, timeout time.Duration) *Blockchain {
+	return &Blockchain{
+		client:    c,
+		bcTimeout: timeout,
+	}
+}
+
+// GetAccountantFee fetches the accountant fee from blockchain
+func (bc *Blockchain) GetAccountantFee(accountantAddress common.Address) (uint16, error) {
+	caller, err := bindings.NewAccountantImplementationCaller(accountantAddress, bc.client)
+	if err != nil {
+		return 0, errors.Wrap(err, "could not create accountant implementation caller")
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), bc.bcTimeout)
+	defer cancel()
+
+	res, err := caller.LastFee(&bind.CallOpts{
+		Context: ctx,
+	})
+	if err != nil {
+		return 0, errors.Wrap(err, "could not get accountant fee")
+	}
+
+	return res.Value, err
+}
+
+// IsRegisteredAsProvider checks if the provider is registered with the accountant properly
+func (bc *Blockchain) IsRegisteredAsProvider(accountantAddress, registryAddress, addressToCheck common.Address) (bool, error) {
+	registered, err := bc.IsRegistered(registryAddress, addressToCheck)
+	if err != nil {
+		return false, errors.Wrap(err, "could not check registration status")
+	}
+
+	if !registered {
+		return false, nil
+	}
+
+	res, err := bc.getProviderChannelLoan(accountantAddress, addressToCheck)
+	if err != nil {
+		return false, errors.Wrap(err, "could not get provider channel loan amount")
+	}
+
+	return res.Cmp(big.NewInt(0)) == 1, nil
```

Comment:
Does that mean at this point that if channel size (loan) is more then 0 we consider it as a consumer?

Reply:
It means that if we have a loan, we're a provider. So if:
1) stake >0: we're a provider
2) stake <= 0: we're a consumer